### PR TITLE
feat(mimo-v2): add EPD disaggregation support

### PR DESCRIPTION
**Pull request by Abatom**
## Summary                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                            
  Add EPD (Encoder–Prefill–Decode) disaggregation support for MiMo-V2, enabling the visual/audio encoder to run on a separate node from the language model. This mirrors the encoder-disaggregation pattern already supported for Qwen2-VL / Qwen3-VL / Qwen2.5-Omni and reuses the existing `encode_server` transport.                                                                                                                                                                             
                                                                                                                                                                                                                                                              
  ## Motivation                                                                                                                                                                                                                                               
   
  MiMo-V2 is a multimodal model whose visual/audio encoders and LLM backbone have very different compute profiles. Co-locating them forces one-size-fits-all TP/DP sizing and wastes GPU memory on idle components. EPD lets operators:                       
                                                                                                                                                                                                                                                            
  - run the encoder on cheaper/less GPUs with its own TP,                                                                                                                                                                                                     
  - run the language model with its preferred TP/DP topology,                                                                                                                                                                                               
  - scale the two tiers independently based on request mix.                                                                                                                                                                                                   
                                                                                                                                                                                                                                                              
  ## Changes by area                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                              
  ### `models/mimo_v2.py` — encoder-only / language-only modes                                                                                                                                                                                                
  - Split `MiMoV2ForCausalLM` construction by `config.encoder_only` / `config.language_only`:                                                                                                                                                               
    - `encoder_only` skips `MiMoV2Model`, `lm_head`, and `LogitsProcessor`.                                                                                                                                                                                   
    - `language_only` skips the vision + audio towers.                                                                                                                                                                                                        
  - New `preprocess_mm_for_encoder(mm_data, modality, config)` hook — lazily builds a `MiMoProcessor` via `from_hf_config` and delegates to `preprocess_for_encoder`. This is the contract the encoder server calls into.                                     
  - New weight-routing helpers (`_LANGUAGE_WEIGHT_PREFIXES`, `_VISION_AUDIO_WEIGHT_PREFIXES`, `_VISION_AUDIO_WEIGHT_SUBSTRING`); `load_weights` now filters by mode so each role only loads its own shards.                                                   
  - Defensive asserts on `forward` / `get_{image,video,audio}_feature` / `get_input_embedding` / `get_embed_and_head` so illegal calls fail fast in the wrong mode.                                                                                           
                                                                                                                                                                                                                                                              
  ### `multimodal/processors/mimo_v2.py` — encoder-side preprocessing + decoder-side reconstruction                                                                                                                                                           
  - `MiMoProcessor.from_hf_config(hf_config, mm_config=None, **overrides)` — builds a minimal, tokenizer-less processor directly from the HF config so the encoder side can run without the tokenizer baggage. Only passes values that come from configs or   
  are required by `__init__`; everything else falls back to `__init__` defaults to avoid silent drift.                                                                                                                                                        
  - `MiMoProcessor.preprocess_for_encoder(mm_data, modality)` — turns raw image/video/audio inputs into the model-ready tensors the single-node path produces (`pixel_values` + `*_grid_thw`, `video_timestamps`, `input_features` +                        
  `audio_feature_lens_raw`).                                                                                                                                                                                                                                  
  - `MiMoProcessor._load_video_frames(video_data)` — shared helper for frame sampling + timestamp generation.                                                                                                                                               
  - `MiMoV2Processor.get_mm_data(prompt, embeddings, **kwargs)` — decoder-side counterpart. Rebuilds `input_ids` with the same per-chunk tokenization as the online path, slices precomputed embeddings into per-placeholder `MultimodalDataItem`s, and       
  computes mrope positions.                                                                                                                                                                                                                                   
  - `_build_mm_token_ids(grid, modality, timestamps=None)` helper for image/video placeholder expansion (with per-frame timestamp tokens when available).                                                                                                     
                                                                                                                                                                                                                                                              
  ### `disaggregation/encode_server.py` — generic model-preprocessor hook                                                                                                                                                                                   
  - Refactor `_process_mm_items` into per-modality helpers: `_process_image_items`, `_process_video_items`, `_process_audio_items`.                                                                                                                           
  - If the loaded model defines `preprocess_mm_for_encoder`, the encoder delegates preprocessing to the model — avoiding the need to add model-specific branches to `encode_server` for every new architecture. Existing Kimi-K2.5 / Kimi-VL image            
  normalization and the `qwen3_5` / `intern_s2_preview` video-timestamp logic are preserved on the fallback path.                                                                                                                                             
  - `get_feature_method` resolution factored out: `getattr(target, f"get_{modality.name.lower()}_feature")` (one line, works through `model.thinker` for omni models).                                                                                        
                                                                                                                                                                                                                                                              
  ### `server_args.py`                                                                                                                                                                                                                                      
  - Skip MiMoV2 fused-QKV TP expectation check when `encoder_only=True` (the language backbone isn't present on that role).                                                                                                                                   
  - Add `MiMoV2ForCausalLM` to the encoder-disaggregation allow-list so `encoder_only` / `language_only` startups don't get rejected.                                                                                                                         
                                                                                                                                                                                                                                                              
  ## Test plan                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                              
  - [ ] Load MiMoV2 with `--encoder-only` on one node, confirm only vision/audio weights load.                                                                                                                                                              
  - [ ] Load MiMoV2 with `--language-only` on another node, confirm only language weights load.                                                                                                                                                               
  - [ ] Run end-to-end EPD request with an image prompt; verify tokens match single-node baseline.                                                                                                                                                            
  - [ ] Same as above with a video prompt (cover timestamp path).                                                                                                                                                                                             
  - [ ] Same as above with an audio prompt.                                                                                                                                                                                                                   
  - [ ] Single-node (non-EPD) MiMoV2 regression check — `preprocess_mm_for_encoder` should be untouched.                                                                                                                                                      
  - [ ] Qwen2-VL / Qwen3-VL / Kimi EPD regression — they don't implement `preprocess_mm_for_encoder`, so the encoder falls back to the refactored per-modality helpers.

### CODE REVIEW

**Review by ShangmingCai — COMMENTED** [review]
Is this PR still a draft? Feel free to ping the reviewer when it is ready.

**Comment by Abatom** [comment]
@ShangmingCai Ready for review!

File: python/sglang/srt/disaggregation/encode_receiver.py [hunk 1]
```diff
@@ -193,7 +193,17 @@ def copy_without_embedding(self):
...
+# Per-part video metadata; keys must match MiMoProcessor.preprocess_for_encoder.
+# Tensor attrs cat on dim=0 across parts; others chain as lists.
+VIDEO_META_ATTRS = (
+    "video_timestamps",
+    "second_per_grid_ts",
+    "video_audio_feature_lens",
+    "video_audio_segment_lens_flat",
+    "video_audio_per_video_num_units",
+    "video_audio_embedding",
+)
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/encode_receiver.py]: Are these new keys MiMo-specific meta?

`keys must match MiMoProcessor.preprocess_for_encoder.` seems a little bit misleading when people are using EPD for other models.

File: python/sglang/srt/disaggregation/encode_server.py [hunk 2]
```diff
@@ -1093,6 +1104,71 @@ async def _encode(self, mm_items, modality: Modality) -> torch.Tensor:
...
+                        if trimmed_chunks
+                        else audio_embedding[:0]
+                    )
+                    aux_data["video_audio_embedding"] = audio_embedding
+                else:
+                    logger.warning(
+                        "Videos carry audio tracks but model has no "
+                        "get_audio_feature; dropping audio for EPD encoding."
+                    )
+
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/encode_server.py]: Should we abstract this part?

File: python/sglang/srt/disaggregation/encode_receiver.py [hunk 3]
```diff
@@ -313,11 +323,8 @@ def get_embedding(self, is_concat=False):
...
             for i, e in enumerate(self.embedding_list):
                 if e is not None:
-                    groups[self.modality_list[i]].append(e.cuda())
-            return {
-                mod: torch.concat(tensors).to("cpu", non_blocking=True)
-                for mod, tensors in groups.items()
-            }
+                    groups[self.modality_list[i]].append(e)
+            return {mod: torch.cat(tensors, dim=0) for mod, tensors in groups.items()}
         return self.embedding_list
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/encode_receiver.py:347]: Will this cause a device mismatch issue?

**Review by ShangmingCai — APPROVED** [review]
Others LGTM. Great PR.

File: python/sglang/srt/disaggregation/encode_receiver.py (hunk 1 above)
Comment (Abatom) [reply python/sglang/srt/disaggregation/encode_receiver.py]: @ShangmingCai Good catch — the previous comment lumped all 6 fields under MiMo, which was misleading. Updated as follows:                                                                                                                                 
                                                                                                                                                                                                                                                
Split into semantic groups:                                                                                                                                                                                                                               
  - _GENERAL_VIDEO_META_ATTRS (video_timestamps, second_per_grid_ts) — shared by both Qwen2/3-VL family and MiMo-VL.                                                                                                                                      
 - _MIMO_VIDEO_AUDIO_META_ATTRS (4 × video_audio_*) — currently emitted only by MiMoProcessor, but the names are model-agnostic and reusable by any future model with an audio-in-video pipeline.                                                          
                                                                                                                                                                                                                                                           
 Rewrote the top-level comment from "keys must match MiMoProcessor" to describe EPD's extension contract — telling future contributors how to onboard a new model (add per-part fields to the appropriate tuple; add to _VIDEO_META_TENSOR_ATTRS if tensor-valued).                                                                                                                                                                                                                                  

No behavior change: VIDEO_META_ATTRS = _GENERAL_VIDEO_META_ATTRS + _MIMO_VIDEO_AUDIO_META_ATTRS preserves the original order and contents, so all 5 iteration sites (4 in encode_receiver.py + encode_server.py:179) are untouched.

File: python/sglang/srt/disaggregation/encode_server.py (hunk 2 above)
Comment (Abatom) [reply python/sglang/srt/disaggregation/encode_server.py]: Refactored into a model-side encode_video_audio hook; server only dispatches now.

File: python/sglang/srt/disaggregation/encode_receiver.py (hunk 3 above)
Comment (Abatom) [reply python/sglang/srt/disaggregation/encode_receiver.py:347]: No — every entry in embedding_list is a CPU tensor: ZMQ path uses torch.frombuffer(...).clone(), Mooncake path allocates torch.empty(total_bytes, dtype=torch.uint8). Neither lands on GPU, so torch.cat over them is same-device by construction.        

The previous .cuda() + torch.concat + .to("cpu", non_blocking) was a perf trick (H2D upload, GPU concat, async D2H), not a correctness fix. Dropped it to keep everything CPU-resident; if profiling shows concat becomes a bottleneck at large batch, we can reintroduce a GPU-concat path with proper benchmarking.

File: python/sglang/srt/disaggregation/encode_server.py [hunk 4]
```diff
@@ -1093,6 +1104,22 @@ async def _encode(self, mm_items, modality: Modality) -> torch.Tensor:
                 self.profiler.step()
 
             aux_data = _build_mm_aux_data(mm_inputs)
+
+            if modality == Modality.VIDEO and mm_inputs.get("video_audio_features"):
+                target = (
+                    self.model.thinker if hasattr(self.model, "thinker") else self.model
+                )
+                hook = getattr(target, "encode_video_audio", None)
+                if hook is not None:
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/encode_server.py]: Give `hook` a better name to reveal the meaning of its functionality

Comment (Abatom) [reply python/sglang/srt/disaggregation/encode_server.py]: Done

**Comment by ZhengWG** [comment]
Great PR! `preprocess_mm_for_encoder` is effectively the new contract. I think, after this lands, it'd be worth opening a tracking issue to migrate the remaining model_type branches in encode_server (Kimi image normalization, Qwen3-VL video_timestamps, Qwen2.5-VL/Omni second_per_grid_ts, etc.) into each model's processor under the same hook. cc @ShangmingCai @liusy58 @gty111

**Review by liusy58 — APPROVED** [review]

**Comment by ShangmingCai** [comment]
CI has passed.
image
